xpti/include/xpti_trace_framework.hpp

```diff
@@ -0,0 +1,367 @@
+            int  ret;
+
+            ret = wcstombs(buffer, &w[0], w.size());
+            // If size is less than PATH_MAX, it is
+            // NULL terminated, else we termintate it
+            // and the file path will be truncated.
+            if(ret == PATH_MAX) {
+                buffer[PATH_MAX-1] = '\0';
+            }
+            else {
+                // However, the string is not
+                // NULL terminated, so we do
+                // it anyway
+                buffer[ret] = '\0';
+            }
+
+            return buffer;
+        }
+#endif
+
+        /// @brief Finds a function defined by symbol in a shared object or DLL
+        /// @details This function is a platform independent abstraction for finding
+        /// a symbol in a shared object of DLL. If successful, it will return a
+        /// valid function pointer.
+        ///
+        /// @param [in]  h        The plugin handle which is required to make this
+        ///                       query
+        /// @param [in] symbol    The symbol that needs to be searched within the
+        ///                       shared object or DLL represented by the handle h
+        ///
+        /// @return <addr>        If the query is successful, the function returns
+        ///                       a valid address
+        /// @return nullptr       If the query fails, nullptr is returned
+        ///
+        xpti_plugin_function_t findFunction(xpti_plugin_handle_t h, const char *symbol) {
+            xpti_plugin_function_t func = nullptr;
+            if (h && symbol) {
+#if defined(_WIN32) || defined(_WIN64)
+                func = GetProcAddress(h, symbol);
+#else
+                func = dlsym(h, symbol);
+#endif
+            }
+            return func;
+        }
+
+        /// @brief Loads a shared object or DLL and returns a plugin handle
+        /// @details This function is a platform independent abstraction for loading
+        /// a DLL or shared object and returns a valid plugin handle if successful.
+        ///
+        /// @param [in]  h        The plugin handle which is required to make this
+        ///                       query
+        /// @param [in] symbol    The symbol that needs to be searched within the
+        ///                       shared object or DLL represented by the handle h
+        ///
+        /// @return <addr>        If the query is successful, the function returns
+        ///                       a valid address
+        /// @return nullptr       If the query fails, nullptr is returned
+        ///
+        xpti_plugin_handle_t loadLibrary(const wchar_t *path, std::string &error) {
+            xpti_plugin_handle_t handle = 0;
+#if defined(_WIN32) || defined(_WIN64)
+            handle = LoadLibraryW(path);
+            if (!handle) {
+                error = getLastError();
+            }
+#else
+            std::wstring str = path;
+            std::string utf8_file = convWChar2MBChar(str);
```

Comment:
Because wstring was used for API signatures and Linux does not have an API to take in wstring :)  Until I have two different implementations with different API signatures, this will exist - it may make sense to do that.